### PR TITLE
UHF-9287: Hakuvahti module and frontend

### PR DESCRIPTION
**Pull request by sundflux**
# [UHF-9287](https://helsinkisolutionoffice.atlassian.net/browse/UHF-9287)

## What was done

Implements frontend to order Hakuvahti subscriptions.

### Install Rekry Drupal:

* Install rekry instance according to regular install instructions.
  * `git checkout UHF-9287-hakuvahti-frontend`
  * `make fresh`
* Checkout [hdbt theme](https://github.com/City-of-Helsinki/drupal-hdbt) under directory /public/themes/contrib/hdbt
* Check out branch `UHF-9287-hakuvahti-frontend-react` for the theme
* Install deps and compile the theme with `npm i && npm run dev`
* Run `make drush-cr`
* Remember to index Elasticsearch results inside Rekry container (`make shell`) by running `drush sapi-rt && drush sapi-i`

### Install Hakuvahti server:

* Install hakuvahti server according to install [instructions](https://github.com/City-of-Helsinki/helfi-hakuvahti)
* RECOMMENDED to run with local nodejs and local mongodb - docker images are mainly for Openshift
* Get ATV secret key from [azure](https://portal.azure.com/#@helsinginkaupunki.onmicrosoft.com/resource/subscriptions/5050c890-3cab-451b-b763-ad55ff3688de/resourceGroups/hki-kanslia-shared-we-devtest-rg/providers/Microsoft.KeyVault/vaults/hki-kxORZUfe-test-kv/secrets) for test environment
* Start the Hakuvahti server with `npm run dev`
* (optional) If MongoDB container doesn't work or connect for some reason, install it locally and connect to mongodb://127.0.0.1:27017/hakuvahti instead of mongodb://mongodb:27017/hakuvahti
* (optional) To quickly install local mongo on macOS: `xcode-select --install && brew tap mongodb/brew && brew update && brew install mongodb-community@7.0 && brew services start mongodb-community@7.0`
* To see and test the emails locally, install + configure [Mailhog](https://github.com/mailhog/MailHog)
* Start Mailhog with: `mailhog -ui-bind-addr localhost:8088 -smtp-bind-addr localhost:1025 -api-bind-addr localhost:8088` to match config below

.env for Hakuvahti server:

```
ENVIRONMENT=dev
FASTIFY_PORT=3000
MONGODB=mongodb://127.0.0.1:27017/hakuvahti
#MONGODB=mongodb://mongodb:27017/hakuvahti
SENTRY_DSN=
ELASTIC_PROXY_URL=https://elastic-helfi-rekry.docker.so
BASE_URL=https://helfi-rekry.docker.so
ATV_API_KEY=
ATV_API_URL=https://atv-api-hki-kanslia-atv-test.agw.arodevtest.hel.fi
SUBSCRIPTION_MAX_AGE=90
UNCONFIRMED_SUBSCRIPTION_MAX_AGE=5
SUBSCRIPTION_EXPIRY_NOTIFICATION_DAYS=3
MAIL_FROM="some@email.fi"
MAIL_HOST=helfi-rekry.docker.so
MAIL_PORT=1025
MAIL_SECURE=false
MAIL_AUTH_USER=test
MAIL_AUTH_PASS=test
MAIL_CONFIRMATION_LINK=https://helfi-rekry.docker.so
MAIL_TEMPLATE_PATH=rekry
REMOVE_CONFIRMATION_LINK=https://helfi-rekry.docker.so/hakuvahti/unsubscribe
```

**Get SENTRY_DSN and ATV_API_KEY from [Azure](https://portal.azure.com/#@helsinginkaupunki.onmicrosoft.com/resource/subscriptions/5050c890-3cab-451b-b763-ad55ff3688de/resourceGroups/hki-kanslia-shared-we-devtest-rg/providers/Microsoft.KeyVault/vaults/hki-kxORZUfe-test-kv/secrets)!** 

## How to test

* After all the steps above are done, you should see "Tilaa hakuvahti" on the job search page.
* To run Hakuvahti server actions (there's no local cron in docker images), in Hakuvahti server directory, run `npm run hav:populate-email-queue && npm run hav:send-emails-in-queue` 
* Check mailhog for new result emails and unsubscribe link.
### Check these:

* [ ] Check that subscription form for Hakuvahti gets sent to server successfully
* [ ] Check that the subscription confirmation works (email confirmation link)
* [ ] Check that unsubscribing link works in results email
* [ ] Check that you get new results for your search query if you add new content matching your search query (remember to reindex elastic results)

## Continuous documentation

* [ ] This feature has been documented/the documentation has been updated
* [ ] This change doesn't require updates to the documentation

## Translations

* This PR has still broken translations. They will be fixed in the UHF-10004 PR.

## Other PRs

* https://github.com/City-of-Helsinki/drupal-helfi-rekry/pull/465
* https://github.com/City-of-Helsinki/drupal-hdbt/pull/987

[UHF-9287]: https://helsinkisolutionoffice.atlassian.net/browse/UHF-9287?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: public/modules/custom/helfi_hakuvahti/src/Controller/HelfiHakuvahtiConfirmController.php [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+   * @param \Symfony\Component\DependencyInjection\ContainerInterface $container
+   *   The container.
+   * @param \Symfony\Component\HttpFoundation\RequestStack $request_stack
+   *   The request stack.
+   * @param \Drupal\Core\Utility\Token $token_service
+   *   The token service.
+   * @param \Drupal\Core\Session\AccountInterface $user
+   *   The user account.
+   */
+  public function __construct(
```
[the remark this answers is not in the document]
Comment (sundflux) [reply public/modules/custom/helfi_hakuvahti/src/Controller/HelfiHakuvahtiConfirmController.php:35]: Fixed

File: public/modules/custom/helfi_hakuvahti/src/Controller/HelfiHakuvahtiConfirmController.php [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+   *   saved search form.
+   */
+  public function __invoke(): array {
+    $build = [];
+
+    $request = $this->requestStack->getCurrentRequest();
+    $hash = $request->query->get('hash');
+    $subscription = $request->query->get('subscription');
+
+    if ($this->isFormSubmitted()) {
```
[the remark this answers is not in the document]
Comment (sundflux) [reply public/modules/custom/helfi_hakuvahti/src/Controller/HelfiHakuvahtiConfirmController.php:65]: Nestings gone

File: public/modules/custom/helfi_hakuvahti/src/Controller/HelfiHakuvahtiUnsubscribeController.php [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+   *   The form
+   */
+  public function __invoke(): array {
+    $build = [];
+
+    $request = $this->requestStack->getCurrentRequest();
+    $hash = $request->query->get('hash');
+    $subscription = $request->query->get('subscription');
+
+    if ($this->isFormSubmitted()) {
```
[the remark this answers is not in the document]
Comment (sundflux) [reply public/modules/custom/helfi_hakuvahti/src/Controller/HelfiHakuvahtiUnsubscribeController.php:66]: Nestings gone

File: public/modules/custom/helfi_hakuvahti/src/Controller/HelfiHakuvahtiConfirmController.php [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+ * Confirms a subscription.
+ */
+final class HelfiHakuvahtiConfirmController extends ControllerBase {
+
+  /**
+   * The http client.
+   *
+   * @var \GuzzleHttp\ClientInterface
+   */
+  protected $httpClient;
```
[the remark this answers is not in the document]
Comment (sundflux) [reply public/modules/custom/helfi_hakuvahti/src/Controller/HelfiHakuvahtiConfirmController.php]: All constructors changed to php8 style

**Review by rpnykanen — APPROVED** [review]

**Review by teroelonen — APPROVED** [review]
The functionality has been tested multiple times and it works 🦖

**Comment by teroelonen** [comment]
The SonarCloud failing because of duplicate code was approved by Lari.

**Pull request merged**